<a href="https://colab.research.google.com/github/matzz-11/Quantum_Colab.ipynb/blob/main/3_Pacote_de_ondas_e_barreira_de_potencial.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#Pacote de ondas e barreira de potencial

---

## Requisitos do Código

Esse código será um pouco diferente, iremos trabalhar com uma **animação**! No entanto, as bibliotecas serão as mesmas dos outros códigos, pois uma delas já contém a parte de animações! São utilizadas para cálculos, plot de gráficos (nesse caso animado), cálculo matricial para a animação, display limpo dos resultados e criação dos sliders interativos, respectivamente:

- numpy
- matplotlib
- scipy
- IPython.display
- ipywidgets

***Atenção! Por se tratar de uma animação, esse código ainda está em desenvolvimento, sendo que atualmente temos poucas opções de parâmetros para não sobrecarregar as animações, mas em breve iremos deixar mais completo e interativo! Obrigado.***

**NÃO SE ESQUEÇA DOS DESAFIOS APÓS O CÓDIGO PARA TREINAR!!**

---

## Potencial e Função de Onda

Dessa vez o potencial que nossa partícula está submetida é nulo! Ou seja,

$$
V(x) = 0
$$

Dessa forma, a equação de Schrodinger fica mais simples:

$$
-\frac{ħ^2}{2m} \frac{d^2𝛷(x)}{dx^2} = E𝛷(x)
$$

Podemos introduzir uma constante para simplificar:

$$
k = \sqrt{\frac{2mE}{ħ}}
$$

Reorganizando:

$$
\frac{d^2𝛷(x)}{dx^2} + k^2𝛷(x) = 0
$$

Essa equação é a mesma que encontramos no Poço de Potencial Infinito, mas agora sem as condições de contorno! Veremos que essa pequena mudança irá alterar significativamente os resultados.

---

## Solução

Novamente, temos a solução geral dada por:

$$
𝛷(x) = Ae^{ikx} + Be^{-ikx}
$$

ou

$$
𝛷(x) = e^{ikx}
$$

*A primeira é a forma mais completa, considerando ondas se propagando para direita e esquerda, "indo e vindo". É mais usada quando queremos descrever fenomênos de reflexão, espalhamento ou simetria. A segunda é um caso particular em que B = 0 e A = 1, ou seja, temos apenas uma onda se propagando para direita. Nesse caso, temos um estado de momento definido, nos permitindo montar pacotes de ondas com A(k). Não se preocupe, essa "montagem" será explicada em breve!*

Não temos condições de contorno para definirmos as constantes A e B! A densidade de probabilidade é constante para todo x, **a partícula tem a mesma probabilidade de estar em qualquer lugar, ou seja, nossa solução não é normalizável!**
Se recorremos ao princípio da Incerteza, teríamos uma situação onde a incerteza na posição x tenderia ao **infinito**, ou seja, a incerteza no momento p tenderia a 0, **teríamos certeza do momento da partícula!**

Para corrigir esse problema, ou seja, torná-lo fisicamente aceitável, montamos uma combinação de ondas com diferentes k (números de onda) através da transformada inversa de Fourier! Dessa forma, geramos pacotes de ondas localizados no espaço, permitindo incertezas finitas tanto na posição quanto no momento.

$$
Ψ(x) = ∫ A(k)e^{ikx} dx
$$

Onde A(k) é a amplitude da contribuição de cada momento!

Curiosamente, a energia de cada partícula se assemelha bastante a **Energia Cinética Clássica!** Basta substituir o caso particular da solução gera (propagação única para direita) na equação de Schrodinger:

$$
-\frac{\hbar^2}{2m} \frac{d^2\phi(x)}{dx^2} = E\phi(x)
$$

Solução:

$$
\phi_k(x) = e^{ikx} \quad \text{com} \quad k = \frac{p}{\hbar}
$$

Substituindo:

$$
\frac{d^2}{dx^2} e^{ikx} = -k^2 e^{ikx}
$$

$$
\Rightarrow -\frac{\hbar^2}{2m} (-k^2)\cdot e^{ikx} = E \cdot e^{ikx}
$$

$$
\Rightarrow \frac{\hbar^2 k^2}{2m} = E
$$

Pela relação de De Broglie:

$$
p = ħk
$$

$$
⇒ E = \frac{p^2}{2m}
$$

Sendo essa a energia de cada partícula livre em função do momento!

---

## Interação com uma barreira de potencial - Caso particular

Agora que sabemos como nosso pacote de onda se comporta segundo a Equação de Schrodinger, conseguimos experimentar o que aconteceria caso ele colidisse com uma barreira de potencial.

De forma intuitiva (não vamos tratar muito da matemática do problema por enquanto), a função de onda precisa existir em todos os locais devido as condições de continuidade da Equação de Schrodinger, ou seja, **ela pode existir atravessando a barreira, mesmo com energia menor**.

O estudo desse fenômeno nos revela duas constantes que serão importantes na simulação, R e T, Coeficiente de Reflexão e Coeficiente de Transmissão, respectivamente. Sua soma deve resultar em 1 (100%), pois a função não pode simplesmente sumir! Dessa forma, pela continuidade, **sempre vamos ter uma parcela de cada um, nunca apenas transmissão ou reflexão, como esperado na mecânica clássica!**

*Apenas uma observação: Como ainda estamos no densenvolvimento do código, pode ser que a soma de R e T nem sempre seja exatamente 1. Fisicamente, isso não é aceitável, mas aqui poderá ocorrer pelo fato de que a simulação está em processo de aprimoramento constante. Pretendemos ajustar isso no futuro!*

---

## Aplicações

Uma das aplicações mais interessantes dos pacotes de ondas livres são os aceleradores de partículas e a Engenharia de partículas. Os pacotes de onda nos ajudam a modelar o movimento, espalhamento e interferência das partículas.
Já para o caso de uma barreira de potencial, temos a **fusão nuclear no Sol**. Para que o Sol "continue funcionando", os prótons em seu centro precisam se unir, mas a energia de repulsão é extremamente alta! Classicamente, eles nunca se uniriam, mas quânticamente vimos que é possível! A continuidade matemática com as condições de contorno nos mostraram que eles podem tunelar por essa barreira, **permitindo a fusão e continuação do processo!**

---

## Conclusão

Para finalizar, temos uma legenda explicando as cores dos itens da animação:

Preto tracejado - Barreira de potencial

Amarelo - Densidade de probabilidade

Azul - Parte real da função de onda

---

##Código


In [ ]:
!pip install -q ipywidgets

# Bibliotecas
import numpy as np
import matplotlib.pyplot as plt
from matplotlib import animation
from scipy.sparse import diags
from scipy.sparse.linalg import splu
from IPython.display import HTML, display
import ipywidgets as widgets

# Constantes para simplificar
hbar = 1.0
m    = 1.0

# Definição da função do pacote de onda
def simulate_wavepacket(a=5.0, V0=2.0, k0=2.0, sigma=2.0,
                        dt=0.02, T_max=100.0, Nx=250, L=250.0,
                        absorb_width=50.0, absorb_strength=1e-2):

# Grade espacial
    x  = np.linspace(0, L, Nx)
    dx = x[1] - x[0]
    center = L/2
    steps = int(T_max / dt)

# Potencial: barreira real + absorção imaginária
    V = np.zeros(Nx, dtype=complex)
    mask_bar = np.abs(x - center) < (a/2)
    V[mask_bar] = V0
    mask_abs = np.zeros(Nx)
    left  = x < absorb_width
    right = x > (L - absorb_width)
    mask_abs[left]  = ((absorb_width - x[left])  / absorb_width)**2
    mask_abs[right] = ((x[right] - (L-absorb_width)) / absorb_width)**2
    V += 1j * absorb_strength * mask_abs

# Pacote inicial normalizado
    x0 = L/4
    env   = np.exp(-((x - x0)**2)/(2*sigma**2))
    phase = np.exp(1j * k0 * x)
    psi0  = env * phase
    psi0 /= np.sqrt(np.sum(np.abs(psi0)**2) * dx)

# Energia do pacote
    E_packet = hbar**2 * k0**2 / (2*m)
    print(f"E_pac = {E_packet:.3f}, V0 = {V0:.3f}")

# Matrizes CN
    coeff    = hbar**2 / (m * dx**2)
    diag_A   = 1 + 1j*dt/(2*hbar)*(coeff + V)
    diag_B   = 1 - 1j*dt/(2*hbar)*(coeff + V)
    off_val  = -1j*dt/(4*hbar)*coeff
    off_arr  = np.full(Nx-1, off_val, dtype=complex)

    A = diags([off_arr, diag_A, off_arr],
              offsets=[-1, 0, 1],
              shape=(Nx, Nx),
              format='csc')
    B = diags([-off_arr, diag_B, -off_arr],
              offsets=[-1, 0, 1],
              shape=(Nx, Nx),
              format='csc')
    A_fac = splu(A)

# Preparação da barreira
    psi = psi0.copy()
    fig, ax = plt.subplots(figsize=(8,4))
    ax.set_xlim(0, L)
    ax.set_ylim(-1.1, 1.1)
    ax.set_xlabel('x')
    ax.set_ylabel(r'$\Re(\Psi)$ & $|\Psi|^2$')
    V_plot = V.real/np.max(V.real)
    ax.plot(x, V_plot, 'k--', label='Barreira')
    raw_line , = ax.plot([], [], label=r'Re[$\Psi$]')
    prob_line, = ax.plot([], [], label=r'$|\Psi|^2$')
    info_text = ax.text(0.02, 0.95, '', transform=ax.transAxes, va='top')
    ax.legend(loc='upper right')

    def init():
        raw_line.set_data([], [])
        prob_line.set_data([], [])
        info_text.set_text('')
        return raw_line, prob_line, info_text

# Amostragem de no máximo 500 quadros para evitar sobrecarga
    max_frames = 500
    frame_idx  = np.linspace(0, steps-1, min(steps, max_frames), dtype=int)

# Definição da função da animação
    def update(i):
        nonlocal psi

# Evolução nos frames
        n_steps = frame_idx[i] - (frame_idx[i-1] if i>0 else -1)
        for _ in range(n_steps):
            psi = A_fac.solve(B.dot(psi))

        t = frame_idx[i] * dt
        raw_line.set_data(x, np.real(psi))
        prob_line.set_data(x, np.abs(psi)**2)
        info_text.set_text(f"t = {t:.2f}")

        if i == len(frame_idx)-1:
            left_area  = x < (center - a/2)
            right_area = x > (center + a/2)
            R = np.sum(np.abs(psi[left_area])**2) * dx
            T = np.sum(np.abs(psi[right_area])**2) * dx
            info_text.set_text(
                f"t={t:.1f} | E={E_packet:.2f}, V0={V0:.2f}, R={R:.3f}, T={T:.3f}"
            )

        return raw_line, prob_line, info_text

    ani = animation.FuncAnimation(
        fig, update, frames=len(frame_idx),
        init_func=init, blit=True, interval=30, repeat=False
    )
    plt.close(fig)
    display(HTML(ani.to_jshtml()))

# Interatividade
widgets.interact_manual(
    simulate_wavepacket,
    a=widgets.FloatSlider(value=5.0, min=1.0, max=5.0, step=1.0,   description='Largura da barreira'),
    V0=widgets.FloatSlider(value=2.0,  min=1.0, max=5.0,   step=1.0, description='Potencial da barreira'),
    k0=widgets.FloatSlider(value=2.0,  min=1.0, max=3.0,  step=1.0, description='Número de onda'),
    sigma=widgets.FloatSlider(value=2.0, min=1.0, max=3.0,  step=1.0, description='Dispersão espacial'),
    dt=widgets.FloatLogSlider(value=0.02, base=10, min=-3, max=-1, step=0.1, description='Intervalo de tempo por frame na animação'),
    T_max=widgets.FloatSlider(value=100.0, min=10.0, max=200.0, step=5.0, description='Tempo máximo da animação'),
    Nx=widgets.IntSlider(value=250, min=200, max=400, step=50, description='Suavização da animação'),
    L=widgets.FloatSlider(value=250.0, min=100.0, max=300.0, step=50.0, description='Largura do eixo x'),
    absorb_width=widgets.FloatSlider(value=50.0,min=10.0, max=100.0, step=5.0, description='Largura de absorção do pacote'),
    absorb_strength=widgets.FloatLogSlider(value=1e-2, base=10, min=-4, max=-1, step=0.1, description='Força de absorção do pacote'),
)


   ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 1.6/1.6 MB 7.6 MB/s eta 0:00:00


interactive(children=(FloatSlider(value=5.0, description='Largura da barreira', max=5.0, min=1.0, step=1.0), F…

<function __main__.simulate_wavepacket(a=5.0, V0=2.0, k0=2.0, sigma=2.0, dt=0.02, T_max=100.0, Nx=250, L=250.0, absorb_width=50.0, absorb_strength=0.01)>

## Desafios para fixação

---

### Exercício 1 — Pacote de onda vs Barreira de potencial

Escolha arbitrariamente alguns valores para a intensidade do potencial (V₀), largura da barreira (a), número de onda (k₀) e dispersão espacial (σ).

> **Atividade:**

> 1. Ao escolher diversos valores para os parâmetros informados, qual relação podemos deduzir entre largura/intensidade do potencial e o tunelamento quântico? E quanto ao número de onda e dispersão espacial?

---

### Exercício 2 — Tunelamento Quântico

Coloque os seguintes valores nos parâmetros:
- a (largura da barreira) = 5.0
- V₀ (potencial) = 2.0
- k₀ (número de onda) = 2.0
- σ (distribuição espacial) = 2.0
- Δt (intervalo temporal entre frames) = 0.0200
- T_max (tempo máximo) = 100.00
- Nx (suavização) = 250
- L (largura do eixo x) = 250.00
- Abs largura (largura da absorção do pacote) = 50.00
- Abs força (força da absorção do pacote) = 0.0100

> **Atividade:**

> 1. Observe que a energia do pacote é igual a energia da barreira. Classicamente, o pacote seria refletido por inteiro, isso acontece quanticamente? Por quê?

---

